pkg/openid4ci/models.go

```diff
@@ -0,0 +1,153 @@
+type JWTSigner interface {
+	// Sign is not yet fully defined. TODO: determine final types to use once an
+	// implementation exists in Aries
+	Sign(clientID, issuer, time, nonce string) ([]byte, error)
+}
+
+// ClientConfig represents the configuration for an OpenID4CI client.
+type ClientConfig struct {
+	// Wallet-hosted URI for callback redirect from issuer, after user authorizes issuer.
+	WalletCallbackURI string
+	// The DID of the wallet user. The requested credential will be bound to this DID.
+	UserDID string
+	// The wallet app instance's OIDC client ID.
+	ClientID string
+	// A handler for signing a JWT for proving possession of userDID.
+	// When this creates a JWT, it should add kid, jwk, or x5c field corresponding to a key bound to UserDID.
+	JWTSigner JWTSigner
+}
+
+// Credential represents a credential in an InitiationRequest.
+// It can be used in one of two mutually-exclusive ways:
+//  1. Setting Format and Types to specify the format of the credential to be requested and further parameters
+//     characterising the type of the credential to be requested.
+//  2. Setting just ID to one of the ID values in one of the Supported Credentials objects in the
+//     credentials_supported Credential Issuer metadata parameter.
+type Credential struct {
+	Format string   `json:"format,omitempty"`
+	Types  []string `json:"types,omitempty"`
+	ID     string   `json:"-"`
+}
+
+// InitiationRequest represents the Issuance Initiation Request object received from an Issuer as defined in
+// https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-5.1.
+type InitiationRequest struct {
+	Issuer            string       `json:"issuer,omitempty"`
+	Credentials       []Credential `json:"credentials,omitempty"`
+	PreAuthorizedCode string       `json:"pre-authorized_code,omitempty"` //nolint: tagliatelle // required by spec
+	UserPinRequired   bool         `json:"user_pin_required,omitempty"`   //nolint: tagliatelle // required by spec
+	OpState           string       `json:"op_state,omitempty"`            //nolint: tagliatelle // required by spec
+}
+
+// IssuerMetadata represents metadata about an issuer as obtained from their .well-known OpenID configuration.
+type IssuerMetadata struct {
+	Issuer                             string `json:"issuer,omitempty"`
+	AuthorizationEndpoint              string `json:"authorizationEndpoint,omitempty"`
+	TokenEndpoint                      string `json:"tokenEndpoint,omitempty"`
+	PushedAuthorizationRequestEndpoint string `json:"pushedAuthorizationRequestEndpoint,omitempty"`
+	RequirePushedAuthorizationRequests bool   `json:"requirePushedAuthorizationRequests,omitempty"`
+	CredentialEndpoint                 string `json:"credentialEndpoint,omitempty"`
+}
+
+// ClientState is a type that is returned from the Client.Authorize method for passing in to the Client.Callback method.
+// It's intended to be treated as an opaque object by the caller.
+type ClientState struct {
```

Comment:
client state was required in JS due to some limitations.

Not required for go bindings.

Reply:
Changed per the latest design we discussed